andi/andi.py

```diff
@@ -307,51 +313,87 @@ class we want to override. If ``recursive_overrides`` is True, then
                     full_final_kwargs=full_final_kwargs,
                     dependency_stack=None,
                     overrides=overrides,
-                    recursive_overrides=recursive_overrides)
+                    recursive_overrides=recursive_overrides,
+                    custom_builder_fn=custom_builder_fn,
+                    )
     return plan
 
 
+CustomBuilder = namedtuple("CustomBuilder", ["result_class_or_fn", "factory"])
+
+
 def _plan(class_or_func: Callable, *,
           is_injectable: Callable[[Callable], bool],
           externally_provided: Callable[[Callable], bool],
           full_final_kwargs,
           dependency_stack=None,
           overrides: Callable[[Callable], Optional[Callable]],
-          recursive_overrides: bool = False
+          recursive_overrides: bool = False,
+          custom_builder_fn: Callable[[Callable], Optional[Callable]] = lambda _: None,
+          custom_builder_result: Optional[Callable] = None
           ) -> Tuple[Plan, List[Tuple]]:
     dependency_stack = dependency_stack or []
     is_root_call = not dependency_stack  # For better code reading
-    plan_od = OrderedDict()  # type: MutableMapping[Callable, KwargsSpec]
+    plan_od = OrderedDict()  # type: MutableMapping[Union[Callable, CustomBuilder], KwargsSpec]
     type_for_arg = KwargsSpec()
 
     if externally_provided(strip_annotated(class_or_func)):
         return Plan([(class_or_func, KwargsSpec())], full_final_kwargs=True), []
 
-    # At this point the class/function must be injectable for non root cases
-    assert is_root_call or is_injectable(strip_annotated(class_or_func))
+    if not custom_builder_result:
+        plan_key = class_or_func  # type: Union[Callable, CustomBuilder]
+    else:
+        plan_key = CustomBuilder(custom_builder_result, class_or_func)
+
+    # At this point the class/function must be injectable or built by a custom builder for non root cases
+    assert is_root_call or is_injectable(strip_annotated(class_or_func)) or custom_builder_result
```

Comment:
very minor: would it make sense to check custom_builder_result first, to save a couple of function calls if it's true?